flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/ExecutorImpl.java

```diff
@@ -98,91 +98,85 @@ public class ExecutorImpl implements Executor {
                             .get();
-            sessionHandle = new SessionHandle(UUID.fromString(response.getSessionHandle()));
-            // register heartbeat service
-            heartbeatScheduler.scheduleAtFixedRate(
-                    () ->
-                            getResponse(
-                                    sendRequest(
-                                            TriggerSessionHeartbeatHeaders.getInstance(),
-                                            new SessionMessageParameters(sessionHandle),
-                                            EmptyRequestBody.getInstance())),
-                    heartbeatInterval,
-                    heartbeatInterval,
-                    TimeUnit.MILLISECONDS);
+            SessionHandle sessionHandle =
+                    new SessionHandle(UUID.fromString(response.getSessionHandle()));
+            return new ExecutorImpl(
+                    restClient,
+                    executor,
+                    gatewayAddress,
+                    connectionVersion,
+                    sessionHandle,
+                    heartbeatInterval);
         } catch (Exception e) {
-            throw new SqlClientException(
-                    String.format("Failed to open session to %s", gatewayAddress), e);
+            executor.shutdownNow();
+            throw new SqlClientException("Failed to create the executor.", e);
         }
     }
 
-    public void closeSession() throws SqlExecutionException {
-        if (sessionHandle == null) {
-            return;
-        }
-        try {
-            CompletableFuture<CloseSessionResponseBody> response =
-                    sendRequest(
-                            CloseSessionHeaders.getInstance(),
-                            new SessionMessageParameters(sessionHandle),
-                            EmptyRequestBody.getInstance());
+    private ExecutorImpl(
+            RestClient restClient,
+            ExecutorService executor,
+            InetSocketAddress gatewayAddress,
+            SqlGatewayRestAPIVersion connectionVersion,
+            SessionHandle sessionHandle,
+            long heartbeatInterval) {
+        this.restClient = restClient;
+        this.executorService = executor;
+        this.gatewayAddress = gatewayAddress;
+        this.sessionHandle = sessionHandle;
+        this.connectionVersion = connectionVersion;
 
-            if (!response.get().getStatus().equals(CLOSE_MESSAGE)) {
-                LOG.warn("The status of close session response isn't {}.", CLOSE_MESSAGE);
-            }
-        } catch (Exception e) {
-            LOG.warn(
-                    String.format(
-                            "Unexpected error occurs when closing session %s.", sessionHandle),
-                    e);
-            // ignore any throwable to keep the cleanup running
-        } finally {
-            sessionHandle = null;
-        }
+        this.heartbeatScheduler = Executors.newSingleThreadScheduledExecutor();
```

Comment:
Can we move this scheduler into create method and use a unified resource mangager, such as 'AutoCloseableRegistry' to register and close all resources